New resource attributes

nciemniak opened the pull request:
Resolves Issue #32 

To consider:
I added a raw sql query in order to get the aggregated group data in a single db call. Ideally I'd have written it out using ActiveRecord methods, but myself and @zwolf couldn't think of a way to do it that wouldn't involve making two separate db calls. The raw sql may be less readable, but besides that there doesn't seem to be a downside to doing it this way. thoughts?

camallen commented on app/models/workflow.rb:
```diff
@@ -6,6 +6,29 @@ class Workflow < ApplicationRecord
 
   def groups
-    transcriptions.group(:group_id).count
+    # aggregate group_id, transcription count, and last update info in one db call
```
I'm trying to unpack this but not getting anywhere. I think the `worklfow.group` method name here (which is very close to `Workflow.group`) is not very descriptive. What is the intention of this 'group' query lookup? 

It looks like the nested sub-select query is grouping the `workflow.translations.group(:group_id)` but also counts the transcription data and selects the most recently updated_at (should this use a limit 1?)

It then joins the sub-select on max_date to find more transcriptions and combine the counts...

sadly i don't follow. A nice description of what is needed here vs just the 'group' wording would help.  

Also can the Join please be explicit (INNER JOIN) vs a default?

FWIW - i'd add comments to this method / query to explain what the statements are trying to achieve. 

Also did anyone run an explain on this query to make sure it was hitting the db indexes / it's a scalable query once data starts filling up the transcriptions / workflows tables?

nciemniak replied:
@camallen , thanks for bringing this up. It's really great to know when code that I've written is hard to read/follow. Lets hash this out over a call when you're back from holiday.

nciemniak commented:
After talking this over with @camallen, decided that the best approach would be to perform the SQL operation with two separate active record queries, for the sake of readability (cause that SQL query is really just a big chonker, not very easy to mentally digest).

We also discussed the possibility of saving the aggregated group data (transcription count, updated at, updated by) in the database to prevent us from needing to perform the aggregation every time we view a workflow. This would require updating the saved group data every time a transcription in that group was updated (this would be done via a worker process), and when a transcription was added. The trade off here would be faster loading for workflow page, but less current data (considering a user may view the workflow page before the worker process has finished updating the group data), and also more development time to build this out.

zwolf commented:
This doesn't really feel a lot more readable to me, but I agree with keeping complexity in ruby rather than SQL. I had to pull down the branch and debug it to fully understand what the intention was here--it seems very odd to use an array of dates in the where clause. Would that duplicate a group in `groups_data` if two transcriptions were updated at the same second?

Also, pretty sure this is three queries (group/count, group/max->dates, where).

This workflows endpoint is applicable to two separate index pages (workflows, groups). So ideally we'd see the front end pulling a list of workflows for the workflow index, then need no additional call for the groups index. The subject sets aren't huge, either, so I think adding sidekiq/redis and backgrounding cache counters could be a later optimization.

zwolf commented on app/models/workflow.rb:
```diff
@@ -5,7 +5,48 @@ class Workflow < ApplicationRecord
+      # take the first/latest update_at transcription
+      transcription = data.first
+      # get the count from the number of transcriptions in the group
+      # Note: this may be sub-optimal if a group can have a large number transcriptions
```
Probably want to remove the editorializing in these two lines and open an issue it's a concern.

nciemniak replied:
Valid! I don't think it's a concern, since we do not expect to be working with large enough amounts of data to run into issues with slow queries. Not sure if it's worth it to make note of this as a potential issue in the future, in the case that the size of our data increases over time, since I wouldn't see this sort of thing as being classified as an actual current issue...

zwolf commented on spec/controllers/workflows_controller_spec.rb:
```diff
@@ -22,9 +22,32 @@
 
     it 'serialized transcription groups' do
-      allow_any_instance_of(Workflow).to receive(:groups).and_return({"FIRST" => 2, "SECOND" => 1})
+      # TODO: the results here look out of sync with the actual group data from the workflow
```
Also want to remove these lines and open an issue if necessary. TODOs are not tracked on this repo.

nciemniak replied:
Ah, heck. I fixed the out-of-sync data, just forgot to remove the comment.
